gnovm/pkg/gnolang/type_check.go

```diff
@@ -0,0 +1,886 @@
+			if debug {
+				debug.Printf("xt is primitiveType: %v, ddt: %v \n", xt, ddt)
+			}
+			// this is special when dt is the declared type of x
+			if !isUntyped(xt) {
+				return fmt.Errorf(
+					"cannot use %s as %s without explicit conversion",
+					xt.String(),
+					ddt.String())
+			} else { // xt untyped, carry on with check below
+				dt = ddt.Base
+			}
+		} else {
+			dt = ddt.Base
+		}
+	}
+
+	// General cases.
+	switch cdt := dt.(type) {
+	case PrimitiveType: // case 1
+		// if xt is untyped, ensure dt is compatible.
+		switch xt {
+		case UntypedBoolType:
+			if dt.Kind() == BoolKind {
+				return nil // ok
+			} else {
+				return fmt.Errorf(
+					"cannot use untyped bool as %s",
+					dt.Kind())
+			}
+		case UntypedStringType:
+			if dt.Kind() == StringKind {
+				return nil // ok
+			} else {
+				return fmt.Errorf(
+					"cannot use untyped string as %s",
+					dt.Kind())
+			}
+		// XXX, this is a loose check, we don't have the context
+		// to check if it is an exact integer, e.g. 1.2 or 1.0(1.0 can be converted to int).
+		// this ensure expr like (a % 1.0) pass check, while
+		// expr like (a % 1.2) panic at ConvertUntypedTo, which is a delayed assertion after const evaluated.
+		// assignable does not guarantee convertible.
+		case UntypedBigdecType:
+			switch dt.Kind() {
+			case IntKind, Int8Kind, Int16Kind, Int32Kind,
+				Int64Kind, UintKind, Uint8Kind, Uint16Kind,
+				Uint32Kind, Uint64Kind, BigdecKind, Float32Kind, Float64Kind:
+				return nil // ok
+			default:
+				panic(fmt.Sprintf(
+					"cannot use untyped Bigdec as %s",
+					dt.Kind()))
+			}
+		case UntypedBigintType:
+			switch dt.Kind() {
+			case IntKind, Int8Kind, Int16Kind, Int32Kind,
+				Int64Kind, UintKind, Uint8Kind, Uint16Kind,
+				Uint32Kind, Uint64Kind, BigintKind, BigdecKind, Float32Kind, Float64Kind: // see 0d0
+				return nil // ok
+			default:
+				return fmt.Errorf(
+					"cannot use untyped Bigint as %s",
+					dt.Kind())
+			}
+		case UntypedRuneType:
+			switch dt.Kind() {
+			case IntKind, Int8Kind, Int16Kind, Int32Kind,
+				Int64Kind, UintKind, Uint8Kind, Uint16Kind,
+				Uint32Kind, Uint64Kind, BigintKind, BigdecKind, Float32Kind, Float64Kind:
+				return nil // ok
+			default:
+				return fmt.Errorf(
+					"cannot use untyped rune as %s",
+					dt.Kind())
+			}
+
+		default:
+			if isUntyped(xt) {
+				panic("unexpected untyped type")
+			}
+			if xt.TypeID() == cdt.TypeID() {
+				return nil // ok
+			}
+		}
+	case *PointerType: // case 4 from here on
+		if pt, ok := xt.(*PointerType); ok {
+			return checkAssignableTo(pt.Elt, cdt.Elt, false)
+		}
+	case *ArrayType:
+		if at, ok := xt.(*ArrayType); ok {
+			if at.Len != cdt.Len {
+				return fmt.Errorf(
+					"cannot use %s as %s",
+					at.String(),
+					cdt.String())
+			}
+			return checkAssignableTo(at.Elt, cdt.Elt, false)
+		}
+	case *SliceType:
+		if st, ok := xt.(*SliceType); ok {
+			return checkAssignableTo(st.Elt, cdt.Elt, false)
+		}
+	case *MapType:
```

Comment:
pretty clear with the rename that checkAssignableTo and checkType should be independent functions.
the original intent (but incorrect implementation) was good, but checking assignability doesn't work here.
if the types are not equal, a map is not assignable.

Reply:
i'm trying a poc now.